### PR TITLE
Introduce Web3 provider handler

### PR DESCRIPTION
**Pull request by xitronix**

### CODE REVIEW

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import {Provider} from 'web3/providers';
+import {render} from 'react-dom';
+import {Web3PickerComponent} from './Web3PickerComponent';
+import {State} from 'reactive-properties';
+import {ULWeb3Provider} from '../../ULWeb3Provider';
+import {ensure, ensureNotNull} from '@universal-login/commons';
+import {Network, getConfigForNetwork} from '../../config';
+import {MetamaskEthereum} from '../../models/metamask';
+
+export interface CustomProvider {
```
Comment (xitronix) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: `WebProviderItem`

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  getProvider: (network: Network) => Provider | MetamaskEthereum;
+  icon: string;
+  name: string;
+}
+
+const universalLoginCustomProvider: CustomProvider = {
```
Comment (xitronix) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: universalLoginProviderItem

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  name: 'UniversalLogin',
+  icon: 'UniversalLogin logo',
+  getProvider: (network) => new ULWeb3Provider(getConfigForNetwork(network)),
+};
+
+const metaMaskProvider: CustomProvider = {
```
Comment (xitronix) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: defaultWeb3ProviderItem

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  currentProviderName?: string = undefined;
+
+  constructor(
+    private onPickProvider?: (providerName: CustomProvider) => void,
+    private network: Network = 'mainnet',
+    private parentSelector?: string,
+  ) {
+    if (window.ethereum) {
+      this.providers = this.providers.concat(metaMaskProvider);
```
Comment (xitronix) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: - ethereum.provider
- suport global variable: web3

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+  }
+
+  getProviders() {
+    return this.providers;
+  }
+
+  setProvider(providerName: string) {
+    ensure(this.providers.some(({name}) => name === providerName), Error, `Provider is not exist. Invalid name: ${providerName}`);
```
Comment (xitronix) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: Error -> InvalidProvider
Can't find provider: ${providerName}

File: universal-login-web3/lib/Web3ProviderFactory.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  icon: 'UniversalLogin logo',
+  create: () => new ULWeb3Provider(getConfigForNetwork('kovan')),
+};
+
+export const metaMaskProviderFactory: Web3ProviderFactory = {
+  name: 'MetaMask',
+  icon: 'MetaMask logo',
+  create: () => {
+    ensureNotNull(window.ethereum, Error, 'MetaMask is not enabled');
+    return window.ethereum;
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/Web3ProviderFactory.ts]: Add support for both `window.ethereum` and `window.web3`

File: universal-login-web3/lib/Web3Strategy.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+export class Web3Strategy implements Provider {
+  web3picker: Web3Picker;
+  currentProvider: Provider;
+
+  constructor(public readonly factories: Web3ProviderFactory[]) {
+    this.web3picker = new Web3Picker(this, factories);
+    this.currentProvider = this.web3picker;
+  }
+
+  send(payload: JsonRPCRequest, callback: Callback<JsonRPCResponse>) {
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/Web3Strategy.ts]: Question: do we need to override some other methods as well?

File: universal-login-web3/lib/Web3ProviderFactory.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  create: () => Provider;
+}
+
+export const universalLoginProviderFactory: Web3ProviderFactory = {
+  name: 'UniversalLogin',
+  icon: 'UniversalLogin logo',
+  create: () => new ULWeb3Provider(getConfigForNetwork('kovan')),
+};
+
+export const metaMaskProviderFactory: Web3ProviderFactory = {
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/Web3ProviderFactory.ts]: rename Web3FromBrowserProviderFactory

File: universal-login-web3/lib/models/metamask.ts [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import {Callback, JsonRPCRequest, JsonRPCResponse} from './rpc';
+import {Provider} from 'web3/providers';
 
-export interface MetamaskEthereum {
+export interface MetamaskEthereum extends Provider {
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/models/metamask.ts]: `BrowserWeb3Provider`

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+export class Web3Picker implements Provider {
+  private isVisible = new State(false);
+  private created = false;
+
+  constructor(
+    private web3Strategy: Web3Strategy,
+    private readonly factories: Web3ProviderFactory[],
+  ) {
+    if (window.ethereum) {
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: Unnecessary

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+      this.factories = this.factories.concat(metaMaskProviderFactory);
+    }
+  }
+
+  setProvider(providerName: string) {
+    ensure(this.factories.some(({name}) => name === providerName), Error, `Provider is not exist. Invalid name: ${providerName}`);
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: Provider does not exist

Comment (marekkirejczyk) [reply universal-login-web3/lib/ui/react/Web3Picker.tsx]: Error -> InvalidProvider

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    this.web3Strategy.currentProvider = this.getCustomProvider(providerName).create();
+    this.isVisible.set(false);
+  }
+
+  private getCustomProvider(name: string): Web3ProviderFactory {
+    const customProvider = this.findCustomProvider(name);
+    ensureNotNull(customProvider, Error, `Provider with given name(${name}) is not exist`);
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: InvalidProvider

Comment (marekkirejczyk) [reply universal-login-web3/lib/ui/react/Web3Picker.tsx]: getProvider(name)

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    return customProvider;
+  }
+
+  private findCustomProvider = (name: string) => this.factories.find((customProvider) => customProvider.name === name);
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: findProvider

File: universal-login-web3/lib/ui/react/Web3PickerComponent.tsx [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  customProviders: Web3ProviderFactory[];
+  setProvider: (providerName: string) => void;
+}
+
+export const Web3PickerComponent = ({customProviders, isVisibleProp, hideModal, setProvider}: Web3PickerComponentProps) => {
+  const isVisible = useProperty(isVisibleProp);
+  // const getOnClick = (name: string) => () => {
+  //   setProvider(name);
+  //   hideModal();
+  // };
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/react/Web3PickerComponent.tsx]: Comment

File: universal-login-web3/lib/ui/react/Web3Picker.tsx [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+import {Web3PickerComponent} from './Web3PickerComponent';
+import {State} from 'reactive-properties';
+import {ensure, ensureNotNull} from '@universal-login/commons';
+import {waitForFalse} from '../../utils';
+import {Web3ProviderFactory, metaMaskProviderFactory} from '../../Web3ProviderFactory';
+import {Web3Strategy} from '../../Web3Strategy';
+import {JsonRPCRequest} from '../../models/rpc';
+
+export class Web3Picker implements Provider {
+  private isVisible = new State(false);
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/react/Web3Picker.tsx]: 👍

File: universal-login-web3/lib/Web3ProviderFactory.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  create: () => Provider;
+}
+
+export const universalLoginProviderFactory: Web3ProviderFactory = {
+  name: 'UniversalLogin',
+  icon: 'UniversalLogin logo',
+  create: () => new ULWeb3Provider(getConfigForNetwork('kovan')),
+};
+
+export const browserWeb3ProviderFactory: Web3ProviderFactory = {
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/Web3ProviderFactory.ts]: One per file?

File: universal-login-web3/lib/services/MetamaskService.ts [hunk 17]
```diff
@@ -1,8 +1,8 @@
-import {MetamaskEthereum} from '../models/metamask';
+import {BrowserWeb3Provider} from '../models/BrowserWeb3Provider';
 import {State} from 'reactive-properties';
 
 export class MetamaskService {
-  metamaskProvider = new State<MetamaskEthereum | undefined>(undefined);
+  metamaskProvider = new State<BrowserWeb3Provider | undefined>(undefined);
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/services/MetamaskService.ts]: `Nullable< BrowserWeb3Provider>`?

File: universal-login-web3/lib/ui/utils/errors.ts [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+export class InvalidProvider extends ValidationFailed {
+  constructor(providerName: string) {
+    super(`Invalid provider: ${providerName}`, 'InvalidProvider');
+    Object.setPrototypeOf(this, InvalidProvider.prototype);
+  }
+}
+export class Web3ProviderNotFound extends ValidationFailed {
+  constructor() {
+    super('Browser Web3 provider not found', 'InvalidProvider');
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/utils/errors.ts]: InvalidProvider ->  Web3ProviderNotFound

File: universal-login-web3/lib/ui/utils/errors.ts [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    Object.setPrototypeOf(this, InvalidProvider.prototype);
```
Comment (marekkirejczyk) [inline universal-login-web3/lib/ui/utils/errors.ts]: InvalidProvider -> Web3ProviderNotFound

**Pull request merged**